src/test/java/io/confluent/examples/streams/ReorderIntegrationTest.java

```diff
@@ -0,0 +1,248 @@
+ * by the timestamp embedded in the message payload.
+ * <p>
+ * Makes sense only on per partition basis.
+ * <p>
+ * Reordering occurs within time windows defined by the
+ *
+ * Note: This example uses lambda expressions and thus works with Java 8+ only.
+ */
+public class ReorderIntegrationTest {
+
+  public static class ReorderTransfomer<K, V>
+      implements Transformer<K, V, KeyValue<K, V>> {
+
+    public interface TimestampExtractor<V> {
+      long getTimestamp(V val);
+    }
+
+    public interface StoreKeyGenerator<K, V> {
+      K getStoreKey(K key, V val);
+    }
+
+    private final String storeName;
+    private final Duration grace;
+    private KeyValueStore<K, V> reorderStore;
+    private ProcessorContext context;
+    private final TimestampExtractor<V> timestampExtractor;
+    private final StoreKeyGenerator<K,V> storeKeyGenerator;
+
+    public ReorderTransfomer(final String storeName, final Duration grace,
+                             final TimestampExtractor<V> timestampExtractor,
+                             final StoreKeyGenerator<K,V> storeKeyGenerator) {
+      this.storeName = storeName;
+      this.grace = grace;
+      this.timestampExtractor = timestampExtractor;
+      this.storeKeyGenerator = storeKeyGenerator;
+    }
+
+    @Override
+    public void init(final ProcessorContext context) {
+      this.reorderStore = context.getStateStore(this.storeName);
+      this.context = context;
+      context.schedule(
+          this.grace,
+          PunctuationType.STREAM_TIME,
+          this::punctuate
+      );
+    }
+
+    /**
+     * 1) read the timestamp from the message value
+     * 2) inserts into a KeyValueStore using (timestamp, message-key) pair
+     * as the key and the message-value as the value, this also provides
+     * de-duplication.
+     *
+     * @param key
+     * @param value
+     * @return
+     */
+    @Override
+    public KeyValue<K, V> transform(final K key, final V value) {
+      final long ts = timestampExtractor.getTimestamp(value);
+      final K storeKey = storeKeyGenerator.getStoreKey(key, value);
+      final V storeValue = reorderStore.get(storeKey);
+      if(storeValue == null) {
+        reorderStore.put(storeKey, value);
+      }
+      return null; 	// null suppresses sending to downstream
+    }
+
+    /**
+     * Scheduled to be called automatically when the period
+     * within which message reordering occurs expires.
+     *
+     * Outputs downstream accumulated records sorted by their timestamp.
+     *
+     * 1) read the store using a ranged fetch from 0 to timestamp - 60'000 (=1 minute)
+     * 2) send the fetched messages in order using context.forward() and deletes
+     * them from the store
+     *
+     * @param timestamp – stream time of the punctuate function call
+     */
+    void punctuate(final long timestamp) {
+      try(KeyValueIterator<K, V> it = reorderStore.all()) {
```

Comment:
The comments state to use `range(from, to)`, but the code uses `all()` since the code deletes everything in the store after forwarding, I think it will continue to work as-is, but you should update the comments.

Reply:
Updated comments